File: src/test/java/com/hileco/cortex/instructions/InstructionsBuilderTest.java
package com.hileco.cortex.instructions;

import com.hileco.cortex.vm.VirtualMachine;
import com.hileco.cortex.vm.Program;
import com.hileco.cortex.vm.ProgramContext;
import com.hileco.cortex.instructions.conditions.EQUALS;
import com.hileco.cortex.instructions.conditions.IS_ZERO;
import com.hileco.cortex.instructions.debug.NOOP;
import com.hileco.cortex.instructions.jumps.JUMP_DESTINATION;
import com.hileco.cortex.instructions.jumps.JUMP_IF;
import com.hileco.cortex.instructions.math.ADD;
import com.hileco.cortex.instructions.stack.DUPLICATE;
import com.hileco.cortex.instructions.stack.PUSH;
import org.junit.Assert;
import org.junit.Test;

import java.math.BigInteger;
import java.util.List;

public class InstructionsBuilderTest {

    @Test
    public void testJump() throws ProgramException {
        var program = new Program(BigInteger.ZERO, List.of(
                new PUSH(new byte[]{123}),
                new PUSH(new byte[]{123}),
                new EQUALS(),
                new PUSH(BigInteger.valueOf(10L).toByteArray()),
                new JUMP_IF(),
                new NOOP(),
                new NOOP(),
                new NOOP(),
                new NOOP(),
                new NOOP(),
                new JUMP_DESTINATION()
        ));
        var programContext = new ProgramContext(program);
        var processContext = new VirtualMachine(programContext);
        var programRunner = new ProgramRunner(processContext);
        programRunner.run();
        Assert.assertEquals(program.getInstructions().size(), programContext.getInstructionPosition());
        Assert.assertEquals(6, programContext.getInstructionsExecuted());
    }

    @Test
    public void testNoJump() throws ProgramException {
        var program = new Program(BigInteger.ZERO, List.of(
                new PUSH(new byte[]{123}),
                new PUSH(new byte[]{124}),
                new EQUALS(),
                new PUSH(BigInteger.valueOf(10L).toByteArray()),
                new JUMP_IF(),
                new NOOP(),
                new NOOP(),
                new NOOP(),
                new NOOP(),
                new NOOP(),
                new JUMP_DESTINATION()
        ));
        var programContext = new ProgramContext(program);
        var processContext = new VirtualMachine(programContext);
        var programRunner = new ProgramRunner(processContext);
        programRunner.run();
        Assert.assertEquals(program.getInstructions().size(), programContext.getInstructionPosition());
        Assert.assertEquals(program.getInstructions().size(), programContext.getInstructionsExecuted());
    }

    @Test
    public void testLoop() throws ProgramException {
        var program = new Program(BigInteger.ZERO, List.of(
                new PUSH(new byte[]{0}),
                new JUMP_DESTINATION(),
                new PUSH(new byte[]{1}),
                new ADD(),
                new DUPLICATE(0),
                new PUSH(new byte[]{1, 0}),
                new EQUALS(),
                new IS_ZERO(),
                new PUSH(BigInteger.valueOf(1L).toByteArray()),
                new JUMP_IF()
        ));
        var programContext = new ProgramContext(program);
        var processContext = new VirtualMachine(programContext);
        var programRunner = new ProgramRunner(processContext);
        programRunner.run();
        Assert.assertEquals(program.getInstructions().size(), programContext.getInstructionPosition());
        Assert.assertEquals(((program.getInstructions().size() - 1) * 256) + 1, programContext.getInstructionsExecuted());
    }

}

File: src/test/java/com/hileco/cortex/instructions/InstructionsBuilderTest.kt
package com.hileco.cortex.instructions

import com.hileco.cortex.instructions.conditions.EQUALS
import com.hileco.cortex.instructions.conditions.IS_ZERO
import com.hileco.cortex.instructions.debug.NOOP
import com.hileco.cortex.instructions.jumps.JUMP_DESTINATION
import com.hileco.cortex.instructions.jumps.JUMP_IF
import com.hileco.cortex.instructions.math.ADD
import com.hileco.cortex.instructions.stack.DUPLICATE
import com.hileco.cortex.instructions.stack.PUSH
import com.hileco.cortex.vm.Program
import com.hileco.cortex.vm.ProgramContext
import com.hileco.cortex.vm.VirtualMachine
import org.junit.Assert
import org.junit.Test

class InstructionsBuilderTest {
    @Test
    fun testJump() {
        val program = Program(listOf(
                PUSH(byteArrayOf(123)),
                PUSH(byteArrayOf(123)),
                EQUALS(),
                PUSH(10),
                JUMP_IF(),
                NOOP(),
                NOOP(),
                NOOP(),
                NOOP(),
                NOOP(),
                JUMP_DESTINATION()
        ))
        val programContext = ProgramContext(program)
        val virtualMachine = VirtualMachine(programContext)
        val programRunner = ProgramRunner(virtualMachine)
        programRunner.run()
        Assert.assertEquals(program.instructions.size.toLong(), programContext.instructionPosition.toLong())
        Assert.assertEquals(6, programContext.instructionsExecuted.toLong())
    }

    @Test
    fun testNoJump() {
        val program = Program(listOf(
                PUSH(byteArrayOf(123)),
                PUSH(byteArrayOf(124)),
                EQUALS(),
                PUSH(10),
                JUMP_IF(),
                NOOP(),
                NOOP(),
                NOOP(),
                NOOP(),
                NOOP(),
                JUMP_DESTINATION()
        ))
        val programContext = ProgramContext(program)
        val virtualMachine = VirtualMachine(programContext)
        val programRunner = ProgramRunner(virtualMachine)
        programRunner.run()
        Assert.assertEquals(program.instructions.size.toLong(), programContext.instructionPosition.toLong())
        Assert.assertEquals(program.instructions.size.toLong(), programContext.instructionsExecuted.toLong())
    }

    @Test
    fun testLoop() {
        val program = Program(listOf(
                PUSH(byteArrayOf(0)),
                JUMP_DESTINATION(),
                PUSH(byteArrayOf(1)),
                ADD(),
                DUPLICATE(0),
                PUSH(byteArrayOf(1, 0)),
                EQUALS(),
                IS_ZERO(),
                PUSH(1),
                JUMP_IF()
        ))
        val programContext = ProgramContext(program)
        val virtualMachine = VirtualMachine(programContext)
        val programRunner = ProgramRunner(virtualMachine)
        programRunner.run()
        Assert.assertEquals(program.instructions.size.toLong(), programContext.instructionPosition.toLong())
        Assert.assertEquals(((program.instructions.size - 1) * 256 + 1).toLong(), programContext.instructionsExecuted.toLong())
    }

}
